### PR TITLE
For #19693 - Display a biometric prompt when a credit card is selected to autofill 

### LINKED ISSUE #19693: Display a biometric prompt when a credit card is selected to autofill
Screen Shot 2021-05-27 at 3 35 10 PM

### PR DESCRIPTION
**Pull request by gabrielluong**
Fixes #19693

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by codrut-topliceanu** [comment]
When canceling the biometrics prompt request the cc number is still written in the editText box. 
https://user-images.githubusercontent.com/60002907/119982944-2c263d00-bfc8-11eb-81f6-5c42d416e7a9.mp4

File: app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt [hunk 1]
```diff
@@ -731,6 +754,61 @@ abstract class BaseBrowserFragment : Fragment(), UserInteractionHandler, Activit
...
+            }// else {
+                // navigateToSavedCards()
+            //}
+        }
+    }
+
+    /**
+     * Shows a pin request prompt. This is only used when BiometricPrompt is unavailable.
+     */
+    @Suppress("Deprecation")
```
Comment (Amejia481) [inline app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt:784]: Why we are suppressing here?

File: app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt [hunk 2]
```diff
@@ -731,6 +754,61 @@ abstract class BaseBrowserFragment : Fragment(), UserInteractionHandler, Activit
...
+
+    /**
+     * Shows a dialog warning about setting up a device lock PIN.
+     */
+    private fun showPinDialogWarning(context: Context) {
+        AlertDialog.Builder(context).apply {
+            setTitle(getString(R.string.credit_cards_warning_dialog_title))
+            setMessage(getString(R.string.credit_cards_warning_dialog_message))
+            setNegativeButton(getString(R.string.credit_cards_warning_dialog_later)) { _: DialogInterface, _ ->
+
```
Comment (Amejia481) [inline app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt:801]: Should we dismiss here?

File: app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt [hunk 3]
```diff
@@ -533,6 +542,29 @@ abstract class BaseBrowserFragment : Fragment(), UserInteractionHandler, Activit
...
+                        resultCode = Activity.RESULT_CANCELED
+                    )
+                },
+                onAuthSuccess = {
+                    promptsFeature.get()!!.onActivityResult(
+                        requestCode = PIN_REQUEST,
+                        data = null,
+                        resultCode = Activity.RESULT_OK
+                    )
+                }
```
Comment (Amejia481) [inline app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt:562]: I thought, we were going to update this, with something similar to:
```kotlin
                onAuthFailure = {
                    promptsFeature.get()?.onBiometricResult( isAuthenticated = false)
                    
                },
                onAuthSuccess = {
                    promptsFeature.get()?.onBiometricResult( isAuthenticated = true)
                }
```

File: app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt [hunk 4]
```diff
@@ -731,6 +766,74 @@ abstract class BaseBrowserFragment : Fragment(), UserInteractionHandler, Activit
...
+            // Warn that the device has not been secured
+            if (context.settings().shouldShowSecurityPinWarning) {
+                showPinDialogWarning(context)
+            } else {
+                promptsFeature.get()!!.onActivityResult(
+                    requestCode = PIN_REQUEST,
+                    data = null,
+                    resultCode = Activity.RESULT_OK
+                )
+            }
```
Comment (Amejia481) [inline app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt:793]: It think we should use instead `promptsFeature.get()?.onBiometricResult( isAuthenticated = true)`

File: app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt [hunk 5]
```diff
@@ -731,6 +766,74 @@ abstract class BaseBrowserFragment : Fragment(), UserInteractionHandler, Activit
...
+        AlertDialog.Builder(context).apply {
+            setTitle(getString(R.string.credit_cards_warning_dialog_title))
+            setMessage(getString(R.string.credit_cards_warning_dialog_message))
+
+            setNegativeButton(getString(R.string.credit_cards_warning_dialog_later)) { _: DialogInterface, _ ->
+                promptsFeature.get()!!.onActivityResult(
+                    requestCode = PIN_REQUEST,
+                    data = null,
+                    resultCode = Activity.RESULT_CANCELED
+                )
```
Comment (Amejia481) [inline app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt:822]: It think we should use instead `promptsFeature.get()?.onBiometricResult( isAuthenticated = false)`

File: app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt [hunk 6]
```diff
@@ -1309,6 +1412,9 @@ abstract class BaseBrowserFragment : Fragment(), UserInteractionHandler, Activit
         private const val REQUEST_CODE_PROMPT_PERMISSIONS = 2
         private const val REQUEST_CODE_APP_PERMISSIONS = 3
 
+        // The PIN request code
+        private const val PIN_REQUEST = 303
```
Comment (Amejia481) [inline app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt:1416]: Could we use the AC version  from `PromptsFeature`?

File: app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt [hunk 7]
```diff
@@ -731,6 +758,66 @@ abstract class BaseBrowserFragment : Fragment(), UserInteractionHandler, Activit
...
+        AlertDialog.Builder(context).apply {
+            setTitle(getString(R.string.credit_cards_warning_dialog_title))
+            setMessage(getString(R.string.credit_cards_warning_dialog_message))
+
+            setNegativeButton(getString(R.string.credit_cards_warning_dialog_later)) { _: DialogInterface, _ ->
+                promptsFeature.get()?.onBiometricResult(isAuthenticated = false)
+            }
+
+            setPositiveButton(getString(R.string.credit_cards_warning_dialog_set_up_now)) { it: DialogInterface, _ ->
+                it.dismiss()
```
Comment (Amejia481) [inline app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt:810]: I think we have to dismiss the prompt request at this point `promptsFeature.get()?.onBiometricResult(isAuthenticated = false)` as we don't get any feedback if the users change their settings.

Comment (Amejia481) [reply app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt:810]: If the users returns and the pages gets refreshed we are going to be in an inconsistent state

Comment (jonalmeida) [reply app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt:810]: Won't the line below do that? `promptsFeature.get()?.onBiometricResult(isAuthenticated = false)`

File: app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt [hunk 8]
```diff
@@ -731,6 +758,66 @@ abstract class BaseBrowserFragment : Fragment(), UserInteractionHandler, Activit
...
+    /**
+     * Shows a pin request prompt. This is only used when BiometricPrompt is unavailable.
+     */
+    @Suppress("Deprecation")
+    private fun showPinVerification(manager: KeyguardManager) {
+        val intent = manager.createConfirmDeviceCredentialIntent(
+            getString(R.string.credit_cards_biometric_prompt_message_pin),
+            getString(R.string.credit_cards_biometric_prompt_unlock_message)
+        )
+        requireActivity().startActivityForResult(intent, PIN_REQUEST)
```
Comment (Amejia481) [inline app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt:794]: 🎉

**Review by Amejia481 — APPROVED** [review]
Tested locally and it works like a charm!

File: app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt (hunk 7 above)
Comment (Amejia481) [reply app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt:810]: Yes, I think I commented before it was added 😄

**Pull request merged**